cmd/restic/lock.go

```diff
@@ -43,7 +43,38 @@ func lockRepository(ctx context.Context, repo restic.Repository, exclusive bool)
 		lockFn = restic.NewExclusiveLock
 	}
 
-	lock, err := lockFn(ctx, repo)
+	var lock *restic.Lock
+	var err error
+
+	retrySleep := 5 * time.Second
+	retryMessagePrinted := false
+	retryTimeout := time.After(gopts.RetryLock)
+
+retryLoop:
+	for {
+		lock, err = lockFn(ctx, repo)
+		if err != nil && restic.IsAlreadyLocked(err) {
+			select {
+			case <-retryTimeout:
+				debug.Log("repo already locked, timeout expired")
+				// Last lock attempt
+				lock, err = lockFn(ctx, repo)
+				break retryLoop
+			default:
+				debug.Log("repo already locked, retrying in %v", retrySleep)
+				if !retryMessagePrinted {
+					if !gopts.JSON {
+						Verbosef("repo already locked, waiting up to %s for the lock\n", gopts.RetryLock)
+					}
+					retryMessagePrinted = true
+				}
+				time.Sleep(retrySleep)
```

Comment:
The loop should also exit if the context is canceled.

Reply:
The `time.Sleep` statement will block until the sleep duration has ended, even when the context was canceled in the meantime. In particular, if the sleep duration is longer than 5 seconds, that could lead to a significant delay (although ctrl-c leads to a call to exit at some point, which should bypass this during normal usage)

Reply:
I had the same thinking that sleep should be blocking. I wrote a test to check we exit on cancel, but I might have made a mistake in it, as it was exiting on time.